sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_asset_utils.py

```diff
@@ -301,96 +295,173 @@ def _get_file_list_content_hash(file_list) -> str:
     for file_path, file_name in sorted(file_list, key=lambda x: str(x[1]).lower()):
-        _hash.update(("#" + file_name + "#").encode())
+        _hash.update(("#" + str(file_name) + "#").encode())
         _hash.update(str(os.path.getsize(file_path)).encode())
-    for file_path, file_name in sorted(file_list, key=lambda x: str(x[1]).lower()):
+    for file_path, _ in sorted(file_list, key=lambda x: str(x[1]).lower()):
         _hash = _get_file_hash(file_path, _hash)
     return str(_hash.hexdigest())
 
 
-def traverse_directory(
-    root: str,
-    files: List[str],
+def get_local_paths(
+    source_path: str,
+    ignore_file: IgnoreFile = IgnoreFile(),
+) -> Tuple[List[str], Dict[str, str]]:
+    """
+    Get the list of local paths to upload and if symlinks are discovered, return a dictionary of the link,target pairs.
+
+    :param source_path: The path to the local file or directory to upload
+    :type source_path: str
+    :param ignore_file: The ignore file to use when determining which files to upload
+    :type ignore_file: IgnoreFile
+    :return: A tuple of the list of local paths to upload and a dictionary of any symlinks and their targets
+    :rtype: Tuple[List[str], Dict[str, str]]
+    """
+    local_paths = []
+    symlink_dict = {}
+    symlinks = []
+
+    def walk_directory_including_symlinks(source: Union[str, os.PathLike]) -> List[Tuple[str, str]]:
+        """
+        Python's os.walk() function doesn't follow symlinks to directories and subdirectories reliably on
+        all platforms. This function modifies it to check for symlinks, follow them if they're directories,
+        and add them to the list of directories to walk.
+
+        :param source: The directory to walk
+        :type source: str
+        :return: A generator of all directories and files in the directory tree.
+        :rtype: Generator[Tuple[str, str]]
+        """
+
+        for root, dirs, files in os.walk(source, topdown=True):
+            for file in files:
+                filename = os.path.join(root, file)
+                if os.path.islink(filename):
+                    # Follow the symlink and add the target to the list of directories to walk
+                    target = os.readlink(filename)
+                    target = os.path.abspath(target)
+                    symlinks.append(file)
+                    if os.path.isdir(target):
+                        dirs.append(target)
+
+                    target = Path(str(target).replace("\\\\?\\", ""))  # Clean paths on Windows with Python 3.10
+                    relative_path = os.path.relpath(target, source)
+
+                    symlink_dict[file] = {
+                        "target file": convert_windows_path_to_unix(relative_path),
+                        "directory": False
+                    }
+                    if os.path.isdir(target):
+                        symlinks.remove(file)
```

Comment:
The `symlinks` array is not used right?

Reply:
Good catch - just removed. In a previous iteration of this method, I was using it for something.